lib/switchs.nix

```diff
@@ -0,0 +1,75 @@
+{ lib }:
+with lib; {
+  /* Emulate a "switch - case" construct,
+   instead of relying on `if then else if ...` */
+
+  /* Usage:
+       ```nix
+       switch-if [
+         if-clause-1
+         ..
+         if-clause-k
+       ] default-out
+       ```
+       where a if-clause has the form `{ cond = b; out = r; }`
+       the first branch such as `b` is true
+
+     Example:
+       ```nix
+       let alphanum = n: switch-if [
+           { cond = n == 0; out = "zero"; }
+           { cond = n == 1; out = "one";  }
+         ] "I do not count beyond one"; in
+       [ (alphanum 0) (alphanum 1) (alphanum 42) ]
+       ```
+      => `[ "zero" "one" "I do not count beyond one" ]`
+
+  */
+  switch-if = c: d: (findFirst (getAttr "cond") {} c).out or d;
+
+  /* Usage:
+       ```nix
+       switch x [
+         simple-clause-1
+         ..
+         simple-clause-k
+       ] default-out
+       ```
+       where a simple-clause has the form `{ case = p; out = r; }`
+       the first branch such as `p x` is true
+       or
+       ```nix
+       switch [ x1 .. xn ] [
+         complex-clause-1
+         ..
+         complex-clause-k
+       ] default-out
+       ```
+       where a complex-clause is either a simple-clause
+       or has the form { cases = [ p1 .. pn ]; out = r; }
+       in which case the first branch such as all `pi xi` are true
+
+       if the variables p are not functions,
+       they are converted to a `equal p`
+       if `out` is missing then `default-out` is taken
+
+     Example:
+       ```nix
+       let test = v1: v2: with versions; switch [ v1 v2 ] [
+           { cases = [ (x: x * 2 < 4) 8 ]; out = "small, 8"; }
+           { case  = [ 42 42 ]; }          # no `out` means `default-out`
+           { case  = l: fold add 0 l > 42; out = "sum over 42"; }
+           { case  = [ 0 1 ];              out = "zero, one"; }
+         ] "weird"; in
+       [ (test 1 8) (test 40 3) (test 0 1) (test 42 42) (test 1 0)]
+       ```
+      => `[ "small, 8" "sum over 42" "zero, one" "weird" "weird" ]`
+  */
+  switch = var: clauses: default: with preds; let
```

Comment:
I think we should only have `switch`, but the magic should be explicit by checking which of those attributes the case attribute set has:

* `cond`: Is a predicate
* `case`: gets converted to `equal x`
* `cases`: gets converted into `builtins.any (equal true) (builtins.map (equal x) cases)` (not sure how to preserve the ability to have predicates here as well though, but maybe we should just sacrifice it here for simplicity's sake)

Some additional ideas:

* There should be a version of `switch` with `var` as its last argument, so it can be curried easily.
* Having a type of condition for boolean expressions would also be nice (maybe named `cond` and the current renamed to `pred`?), so expressions like `!lib.inNixShell` could be incooperated nicely.

Reply:
@sternenseemann  Don't you like the fact that when some `x` is not a function it gets converted automatically to `equal x`? This looks harmless to me as it would not make sense to compare two functions, and it allows to combine both styles within `cases` (as in `{cases = [somePred x]; ...}` where `somePred` is really a predicate and `x` is not), which I think is a common usecase. Moreover it also frees `cond` for boolean conditions in order to subsume `switchIf`. I like your `oneof` suggesstion and would also add `allof` for the sake of dualization.

Indeed `switch` is hard to flip... maybe we could have `switchFun = cs: dft: x: switch x cs dflt`?

Reply:
Yeah, the “magic” is pretty cool, but I'm also a bit cautious in this regard. There is always the danger of for example a function that is partially applied by accident triggering the magic although you didn't want to — resulting in a weird and hard to debug eval error.

Not sure what the best call is here.